### PR TITLE
Fixes #4871 HTTP 500 on creating an event

### LINKED ISSUE #4871: HTTP 500 on creating an event
**Describe the bug**

The server  returns HTTP 500 when creating or patching an event.

```
{
  "data": {
    "attributes": {
      "original-image-url": "https://media.licdn.com/media/AAEAAQAAAAAAAAdcAAAAJDNkMjFlMTNmLTNjMTMtNGRmNS04OTUwLTEwMzc2NzNhZDIxOA.png",
      "logo-url": "https://events.google.com/io2016/images/touch/homescreen192.png"
    },
    "id": "{{event}}",
    "type": "event"
  }
}
```

**To Reproduce**
1. Make a PATCH request with the data above.

**Expected behavior**

It should be successfully created or updated.

**Stacktrace**

```
Traceback (most recent call last):
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/app.py", line 2309, in __call__
    return self.wsgi_app(environ, start_response)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/app/__init__.py", line 64, in __call__
    return self.app(environ, start_response)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/app.py", line 2295, in wsgi_app
    response = self.handle_exception(e)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/app.py", line 1741, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/src/flask-rest-jsonapi/flask_rest_jsonapi/api.py", line 133, in decorated
    return self.check_permissions(view, view_args, view_kwargs, *args, **kwargs)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/app/api/helpers/permission_manager.py", line 431, in permission_manager
    return permissions[args[0]](view, view_args, view_kwargs, *args, **kwargs)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/app/api/helpers/permissions.py", line 45, in decorator
    return fn(*args, **kwargs)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/app/api/helpers/permission_manager.py", line 61, in is_coorganizer
    return view(*view_args, **view_kwargs)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/src/flask-rest-jsonapi/flask_rest_jsonapi/decorators.py", line 32, in wrapper
    return func(*args, **kwargs)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/site-packages/flask/views.py", line 88, in view
    return self.dispatch_request(*args, **kwargs)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/src/flask-rest-jsonapi/flask_rest_jsonapi/resource.py", line 68, in dispatch_request
    response = method(*args, **kwargs)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/src/flask-rest-jsonapi/flask_rest_jsonapi/decorators.py", line 56, in wrapper
    return func(*args, **kwargs)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/src/flask-rest-jsonapi/flask_rest_jsonapi/resource.py", line 300, in patch
    self._data_layer.update_object(obj, data, kwargs)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/src/flask-rest-jsonapi/flask_rest_jsonapi/data_layers/alchemy.py", line 126, in update_object
    self.before_update_object(obj, data, view_kwargs)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/app/api/events.py", line 423, in before_update_object
    uploaded_images = create_save_image_sizes(data['original_image_url'], 'event', event.id)
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/app/api/helpers/files.py", line 166, in create_save_image_sizes
    'original_image_url': create_save_resized_image(image_file, 0, 0, 0, original_upload_path, resize=False),
  File "/media/srv_twry/work/Projects/Community-Projects/Fossasia/open-event-server/app/api/helpers/files.py", line 81, in create_save_resized_image
    data = urllib.request.urlopen(image_file).read()
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/home/srv_twry/anaconda3/envs/open-event-server/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 403: Forbidden
```

### PR DESCRIPTION
**Pull request by bhaveshAn**
Fixes #4871 HTTP 500 on creating an event

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:

Fixes #4871 HTTP 500 on creating an event

#### Changes proposed in this pull request:

- Add exception handling for HTTPError.

### CODE REVIEW

File: app/api/helpers/files.py [hunk 1]
```diff
@@ -78,13 +79,17 @@ def create_save_resized_image(image_file, basewidth=None, maintain_aspect=None,
         return None
 
     filename = '{filename}.{ext}'.format(filename=get_file_name(), ext=ext)
-    data = urllib.request.urlopen(image_file).read()
+    try:
+        data = urllib.request.urlopen(image_file).read()
+    except urllib.error.HTTPError:
+        raise UnprocessableEntity(
+            {'source': 'attributes/original-image-url'}, 'Invalid Image URL'
```
Comment (srv-twry) [inline app/api/helpers/files.py]: This is a helper file. It can't be so specific. We don't know if the attribute will even contain a `original-image-url`. Please raise a HTTPError and add this specific try except block where we are calling this function.

Comment (bhaveshAn) [reply app/api/helpers/files.py]: > We don't know if the attribute will even contain a original-image-url

We are building `original_image_url` `large_image_url` `thumbnail_image_url` & `icon_image_url`  from `data['original_image_url']` just resizing the same image accordingly.
See https://github.com/fossasia/open-event-server/blob/development/app/api/events.py#LC397
So, it should be kept in helper function only.

File: app/api/helpers/files.py [hunk 2]
```diff
@@ -78,13 +79,17 @@ def create_save_resized_image(image_file, basewidth=None, maintain_aspect=None,
         return None
 
     filename = '{filename}.{ext}'.format(filename=get_file_name(), ext=ext)
-    data = urllib.request.urlopen(image_file).read()
+    try:
+        data = urllib.request.urlopen(image_file).read()
+    except (urllib.error.HTTPError, urllib.error.URLError) as e:
+        raise UnprocessableEntity(
+            {'source': 'attributes/original-image-url'}, 'Invalid Image URL'
```
Comment (srv-twry) [inline app/api/helpers/files.py]: Even though we are using it from only one position, if we want to have such specific error messages then add this method in that class only. Adding it in a helper class means we can't use it elsewhere in the project. we may want to use it in the future from a different location as well. Let's follow the best practices.

File: app/api/events.py [hunk 3]
```diff
@@ -120,6 +121,12 @@ def after_create_object(self, event, data, view_kwargs):
                                  status='accepted')
         save_to_db(role_invite, 'Organiser Role Invite Added')
         if data.get('original_image_url'):
+            try:
+                img = urllib.request.urlopen(data['original_image_url']).read()
```
Comment (Kreijstal) [inline app/api/events.py]: Normally, I wouldn't say anything, but a try catch block of an image requestal should have it's own function if it's going to be repeated more than once. Encapsulate the try catch block in a function of it's own.

**Comment by bhaveshAn** [comment]
> DRY, you're repeating the try catch block.
Also fix Codacy issues.

@Kreijstal Added try catch in separate function and fixed codacy issues.
Please review :)

File: app/api/helpers/files.py [hunk 4]
```diff
@@ -84,7 +84,6 @@ def create_save_resized_image(image_file, basewidth=None, maintain_aspect=None,
         im = Image.open(image_file)
     except IOError:
         raise IOError("Corrupt/Invalid Image")
-
```
Comment (srv-twry) [inline app/api/helpers/files.py]: 1. Unnecessary change. 
2. We need that space in order to group the logic. 
please add it again. Moreover please squash your commits.

Comment (bhaveshAn) [reply app/api/helpers/files.py]: Squash commits while merging.

File: app/api/events.py [hunk 5]
```diff
@@ -394,6 +405,7 @@ def before_update_object(self, event, data, view_kwargs):
         :return:
         """
         if data.get('original_image_url') and data['original_image_url'] != event.original_image_url:
+            validate_original_image_url(data['original_image_url'])
```
Comment (iamareebjamal) [inline app/api/events.py]: I think this method will create 2 requests which is not a great way to handle image loads as images can be quite large.

Comment (srv-twry) [reply app/api/events.py]: Yes, I agree. 
@bhaveshAn It will make two network calls for the same image:
1. To verify the image.
2. In the `create_save_image_sizes` method.

I think what @Kreijstal suggested is sub optimal. Here is my suggestion(same as before):
1. In the `create_save_image_sizes` method just raise the HTTPError/URLError
2. Add a try except block for this method call
``` Python
uploaded_images = create_save_image_sizes(data['original_image_url'], 'event', event.id)
```

Comment (Kreijstal) [reply app/api/events.py]: What would the except method do other than changing the error message, why not let the original error bubble up?

Comment (srv-twry) [reply app/api/events.py]: I only have one issue. The `create_save_image_sizes` is a helper method that can be used from else where in the project as well(May be in the future). It's against the best practices if any helper method expects itself to be called from a specific part of the code. If we add 
``` Python
raise UnprocessableEntity(
            {'source': 'attributes/original-image-url'}, 'Invalid Image URL'
```
in the helper method, then it suggests that we are expecting the data to be called from the event only. How can a helper method be sure that the attribute that has been passed is coming from ```attributes/original-image-url```. Ideally it shouldn't assume anything. 

@iamareebjamal What do you think ?

**Comment by srv-twry** [comment]
@bhaveshAn Please update this PR as soon as you can. Creating an event is something that is required by each and every client project and hence it can be a blocker for them. Moreover only a small change is required to get it merged. :)

File: app/api/events.py (hunk 5 above)
Comment (bhaveshAn) [reply app/api/events.py]: Is this what you want finally

```
raise UnprocessableEntity(
            {'source': ''}, 'Invalid Image URL'
```

Comment (srv-twry) [reply app/api/events.py]: I am suggesting what I had suggested you before. You had even updated it in this commit: https://github.com/fossasia/open-event-server/pull/4872/commits/5cd819f8d712a9df11d2284b0a442a171221f0b1

**Comment by srv-twry** [comment]
> Please update this PR as soon as you can. Creating an event is something that is required by each and every client project and hence it can be a blocker for them. Moreover only a small change is required to get it merged. :)

@bhaveshAn Gentle Reminder. HTTP 500 in case of an event creation isn't good. It should be fixed ASAP. If you're busy or unable to fix it, please pass on the responsibility so that it could be resolved ASAP. It can be a blocker for the client projects. Thanks.

**Comment by bhaveshAn** [comment]
had troubles while pushing new commit in this branch hence closing and making new one.

File: app/api/events.py [hunk 6]
```diff
@@ -132,6 +133,12 @@ def after_create_object(self, event, data, view_kwargs):
                                  status='accepted')
         save_to_db(role_invite, 'Organiser Role Invite Added')
         if data.get('original_image_url'):
+            try:
+                image_file = urllib.request.urlopen(data['original_image_url']).read()
+            except (urllib.error.HTTPError, urllib.error.URLError):
+                raise UnprocessableEntity(
+                    {'source': 'attributes/original-image-url'}, 'Invalid Image URL'
+                )
```
Comment (iamareebjamal) [inline app/api/events.py]: This just moved the util method code inline. How does this fix the issue? This is even worse than before

This even removed the comment containing important discussion as to what was to be done.

Please iterate what your follow up commit would be before sending it so that everyone is on the same page

**Comment by iamareebjamal** [comment]
@bhaveshAn Please don't close a PR like that, this makes us lose important discussion on issues. Tell which issues you are having while pushing commits and we'll try to solve it

**Comment by bhaveshAn** [comment]
Tried a lot but this PR has commit 5cd819f which was there in https://github.com/fossasia/open-event-server/pull/4836
```
rebase in progress; onto 2c9d7ad
You are currently rebasing branch 'issue-4871' on '2c9d7ad'.
  (all conflicts fixed: run "git rebase --continue")
```

**Comment by srv-twry** [comment]
@bhaveshAn Can you explain the issue a bit more.

**Comment by srv-twry** [comment]
Run these commands first and see if you face any problems:
```
git fetch upstream
git rebase upstream/development
git push -f origin branch_name
```

Now post the error message or the screenshot of the error that you get after you run these commands.

**Comment by iamareebjamal** [comment]
As the message is suggesting, run `git rebase --continue`

**Comment by bhaveshAn** [comment]
@srv-twry @iamareebjamal 
screenshot from 2018-06-17 17-40-31

**Comment by iamareebjamal** [comment]
git rebase --abort

**Comment by iamareebjamal** [comment]
Because you have conflicts, fix those and add the files, commit and push

This is why is it recommended to always squash and push. So that rebases are easy

**Comment by srv-twry** [comment]
@bhaveshAn In the second image, after running `git rebase upstream/development` it is saying that you have merge conflicts. 
It is in the `app/api/events.py` file as depicted by the line: `CONFLICT(content) : Merge ...`

Open that file, resolve the merge conflict, and then run `git rebase --continue`. If you get something like that again(could happen since you have multiple commits) then keep resolving conflicts and running `git rebase --continue`. 
then run `git add` once it is successful, then squash your commits and push.

**Comment by bhaveshAn** [comment]
thanks @srv-twry @iamareebjamal

**Comment by Kreijstal** [comment]
This pull request.. just changes the order of the try catches but, it doesn't do anything other than that, does it?

**Comment by srv-twry** [comment]
> This pull request.. just changes the order of the try catches but, it doesn't do anything other than that, does it?

It wraps the call to the method `get_cre...` inside a try except block and hence allows us to catch the HTTPError in case the image url is invalid.

**Comment by bhaveshAn** [comment]
It solves the HTTP 500 on event creation but doesn't use more requests to check whether image is valid or not.

**Comment by Kreijstal** [comment]
Yes but the method already had a try catch on it

**Comment by srv-twry** [comment]
It was for incorrect image i.e. a corrupt one, not the incorrect image url. They are two different exceptions.

**Comment by Kreijstal** [comment]
Ah, well, the exceptions are the same (They have the same error message), but I do notice now that the method is different now.. alright then
